monetio/obs/airnow.py

```diff
@@ -55,101 +75,116 @@ def build_urls(dates, *, daily=False):
 
     urls = []
     fnames = []
-    print("Building AIRNOW URLs...")
-    base_url = "https://s3-us-west-1.amazonaws.com//files.airnowtech.org/airnow/"  # TODO: use S3 URL scheme instead?
+    print("Building AirNow URLs...")
+    if sys.version_info < (3, 7):
+        base_url = "https://s3-us-west-1.amazonaws.com/files.airnowtech.org/airnow/"
+    else:
+        base_url = "s3://files.airnowtech.org/airnow/"
     for dt in dates:
         if daily:
             fname = "daily_data.dat"
         else:
             fname = dt.strftime(r"HourlyData_%Y%m%d%H.dat")
-        # 2017/20170131/HourlyData_2017012408.dat
+        # e.g.
+        # https://s3-us-west-1.amazonaws.com//files.airnowtech.org/airnow/2017/20170108/HourlyData_2016121506.dat
         url = base_url + dt.strftime(r"%Y/%Y%m%d/") + fname
         urls.append(url)
         fnames.append(fname)
-    # https://s3-us-west-1.amazonaws.com//files.airnowtech.org/airnow/2017/20170108/HourlyData_2016121506.dat
-    # or https://files.airnowtech.org/?prefix=airnow/2017/20170108/
 
     # Note: files needed for comparison
     urls = pd.Series(urls, index=None)
     fnames = pd.Series(fnames, index=None)
+
     return urls, fnames
 
 
-def read_csv(fn):
-    """Short summary.
+def read_csv(fn, *, daily=None):
+    """Read an AirNow CSV file.
 
     Parameters
     ----------
-    fn : string
-        file name to read
+    fn
+        File to read, passed to :func:`pandas.read_csv`.
+    daily : bool, optional
+        Is this is a daily (``True``) or hourly (``False``) file?
+        By default, attempt to determine based on the file name.
 
     Returns
     -------
-    type
-        Description of returned object.
-
+    pandas.DataFrame
+        AirNow data in long format without site metadata.
+        Additional processing done by :func:`aggregate_files` / :func:`add_data`
+        not applied.
     """
-    hourly_cols = [
-        "date",
-        "time",
-        "siteid",
-        "site",
-        "utcoffset",
-        "variable",
-        "units",
-        "obs",
-        "source",
-    ]
-    daily_cols = ["date", "siteid", "site", "variable", "units", "obs", "hours", "source"]
-    try:
-        dft = pd.read_csv(
-            fn,
-            delimiter="|",
-            header=None,
+    from monetio.util import _get_pandas_version
+
+    pd_ver = _get_pandas_version()
+
+    if daily is None:
+        if isinstance(fn, Path):
+            fn_str = fn.as_posix()
+        else:
+            fn_str = str(fn)
+
+        if fn_str.endswith("daily_data.dat"):
+            daily = True
+        elif re.search(r"HourlyData_[0-9]{10}.dat", fn_str) is not None:
+            daily = False
+        else:
+            raise ValueError("Could not determine if file is daily or hourly")
+
+    dtype = {"siteid": str, "obs": float}
```

Comment:
Need to ensure MSA code and such are read as str, not int/float. `cmsa_name` and `msa_code` seem to need this or they get convert to float64 (in order to represent missing with NaN).